jython-3-mvp.md

```diff
@@ -0,0 +1,236 @@
+---
+title: Jython 3 Features and MVP
+---
+# Jython 3 Features and a Viable Minimum
+
+This is a discussion document that attempts to describe,
+and to some extent prioritise,
+features for Jython 3,
+based on suggestions collated from various voices on jython-dev
+and in off-list e-mail.
+
+
+## Positioning
+
+We think people will continue to adopt and use Jython if Jython 3 ...
+
+* is a modern version of Python, close to standard in its features.
+* runs on a Java platform that is supported in the long-term.
+* integrates cleanly with Java for access to JDK and user libraries.
+* offers correct concurrency (effectively utilising available CPUs).
+* allows code developed on CPython 3.x to run on Jython 3.x.
+* is well-tested for release and supported for bugs.
+
+The minimum viable product (MVP) is to approach *all* these targets closely,
+that is, it isn't viable if it falls a long way short of any one of them.
+This does not preclude the availability of immature alpha versions.
+(It's a public project, so it is there for anyone to build.)
+
+
+## Version and Forms
+
+* MVP: A Python 3.8 (?) close to standard in ability to run Python.
+  * Only essential differences (e.g. around GC, atomicity).
+  * The full syntax (plus some Java twists e.g. to import from Java).
+  * Close to the whole standard library (where use is not C-specific).
+
+* Builds for Gradle/Maven ecosystem.
+  * MVP: Slim JAR without bundled dependencies.
+  * MVP: Centrally available to cite as a dependency.
+
+* Provides an executable command
+  * MVP: `jython3` command installable on each major OS
+  * MVP: subset of commonly-used python3 command options
+  * Option-compatible with `python3` (MVP: a subset)
+  * Options specific to Jython (JVM options and others)
+
+
+## Performance
+
+* MVP: speed comparable with CPython (say 2x either way).
```

Comment:
I agree in general for MVP. Long-term, I'd like scientific computing to be a proposition without C extensions, if we can manage it.

By some measures, Jython 2 is somewhat faster than CPython, although start-up definitely gives the opposite impression. Full use of the JVM's dynamic language features in JRuby had a massive effect on performance. In my toy, and almost certainly in Jython 3, the byte code interpreter uses `MethodHandle` but not `CallSite`. Generated JVM code would use ` invokedynamic` call sites fully, and I'm hoping for something good.

There's a real danger in setting any kind of target that we spend too much time micro-optimising the byte-code interpreter, when a JVM language is the goal. At the same time, if the basic operations of the abstract object API are much slower than in CPython, we're probably laying the wrong foundation. I could be talked down from the particular numbers.

In some simple loops and calls in my toy version I observe about 40% the speed of CPython. It's a bit disappointing, considering how succinct the Java is, but I'm resisting the temptation to delve further.

Reply:
> Long-term, I'd like scientific computing to be a proposition without C extensions, if we can manage it.

If you talk about Java-ports of NumPy and SciPy, I estimate each one is a project comparable in complexity to Jython itself. In addition one still has to link blas and lapack, which inherently require C glue code. This can be based on existing projects like jblas, but then again, it still uses platform-dependent C code, just written by others. Project Panama may solve this but AFAIK it has not even a planned release date.